app/jobs/course/assessment/answer/auto_grading_job.rb

```diff
@@ -41,17 +43,30 @@ class Course::Assessment::Answer::AutoGradingJob < ApplicationJob
   # @param [String] redirect_to_path The path to redirect when job finishes.
   def perform_tracked(answer, redirect_to_path = nil)
     ActsAsTenant.without_tenant do
+      raise PriorityShouldBeLoweredError if queue_name == 'highest' && answer.question.is_low_priority
+
+      start_time = Time.now
       Course::Assessment::Answer::AutoGradingService.grade(answer)
       if update_exp?(answer.submission)
         Course::Assessment::Submission::CalculateExpService.update_exp(answer.submission)
       end
+
+      if queue_name == 'highest' && (Time.now - start_time > answer.auto_grading_time_limit)
+        downgrade_question_priority(answer)
+      end
     end
 
     redirect_to redirect_to_path
   end
 
   private
 
+  def downgrade_question_priority(answer)
+    question = answer.question
+    question.is_low_priority = true
+    question.save!
+  end
```

Comment:
This should go into the question model method, so you can just do: `answer.question.set_is_low_priority!` in L55.

And you also won't have the function appear in both job files.

Reply:
done

Reply:
Don't move the logic for checking overtime into the model, otherwise your logic will become fragmented. Checking overtime is a job concern.  
Model should just do the updating, which is this function.